lib/presentation/widget/shared_space_details/add_shared_space_member/add_shared_space_member_viewmodel.dart

```diff
@@ -0,0 +1,125 @@
+import 'package:linshare_flutter_app/presentation/redux/states/app_state.dart';
+import 'package:linshare_flutter_app/presentation/util/router/app_navigation.dart';
+import 'package:linshare_flutter_app/presentation/widget/base/base_viewmodel.dart';
+import 'package:redux/redux.dart';
+
+class AddSharedSpaceMemberViewModel extends BaseViewModel {
+  final AppNavigation _appNavigation;
+  final GetAutoCompleteSharingInteractor _getAutoCompleteSharingInteractor;
+  final AddSharedSpaceMemberInteractor _addSharedSpaceMemberInteractor;
+  final GetAllSharedSpaceRolesInteractor _getAllSharedSpaceRolesInteractor;
+  final GetAllSharedSpaceMembersInteractor _getAllSharedSpaceMembersInteractor;
+
+  SharedSpaceRoleName selectedRole = SharedSpaceRoleName.READER;
+
+  AddSharedSpaceMemberViewModel(
+      Store<AppState> store,
+      this._appNavigation,
+      this._getAutoCompleteSharingInteractor,
+      this._addSharedSpaceMemberInteractor,
+      this._getAllSharedSpaceRolesInteractor,
+      this._getAllSharedSpaceMembersInteractor)
+      : super(store);
+
+  void initState() {
+    store.dispatch(_getSharedSpaceRolesAction());
+  }
+
+  void backToSharedSpacesDetails() {
+    _appNavigation.popBack();
+  }
+
+  void selectRole(SharedSpaceRoleName role) {
+    selectedRole = role;
+    _appNavigation.popBack();
+  }
+
+  Future<List<SharedSpaceMember>> getSharedSpaceMembers(SharedSpaceId sharedSpaceId) async {
+    return (await _getAllSharedSpaceMembersInteractor.execute(sharedSpaceId))
+        .map((result) => result is SharedSpaceMembersViewState ? result.members : null)
+        .getOrElse(() => []);
+  }
+
+  Future<List<AutoCompleteResult>> getAutoCompleteSharing(
+      String pattern, SharedSpaceNodeNested sharedSpace, List<SharedSpaceMember> members) async {
+    return await _getAutoCompleteSharingInteractor
+        .execute(AutoCompletePattern(pattern), AutoCompleteType.THREAD_MEMBERS,
+            threadId: ThreadId(sharedSpace.sharedSpaceId.uuid))
+        .then(
+          (viewState) => viewState.fold(
+            (failure) => <AutoCompleteResult>[],
+            (success) => (success as AutoCompleteViewState)
+                .results
+                .where((user) => !members
+                    .map((member) => member.account.mail)
+                    .contains(user.getSuggestionMail()))
+                .toList(),
+          ),
+        );
+  }
+
+  void addSharedSpaceMember(
+      SharedSpaceId sharedSpaceId, SharedSpaceMemberAutoCompleteResult autoCompleteResult) {
+    store.dispatch(
+        _addSharedSpaceAction(sharedSpaceId, AccountId(autoCompleteResult.userUuid.uuid)));
+  }
+
+  OnlineThunkAction _addSharedSpaceAction(SharedSpaceId sharedSpaceId, AccountId accountId) {
+    return OnlineThunkAction((Store<AppState> store) async {
+      final selectedRoleId = store.state.sharedSpaceState.rolesList
+          .firstWhere((element) => element.name == selectedRole)
+          .sharedSpaceRoleId;
```

Comment:
how about the case we can not get `roleIds`, this `selectedRoleId` is null?

Reply:
WDYT?
```dart
  OnlineThunkAction _addSharedSpaceAction(SharedSpaceId sharedSpaceId, AccountId accountId) {
    return OnlineThunkAction((Store<AppState> store) async {
      final rolesList = store.state.sharedSpaceState.rolesList;
      if (rolesList.isEmpty) {
        return;
      }

      final role = rolesList.firstWhere((element) => element.name == selectedRole, orElse: () => null);
      if (role == null) {
        return;
      }

      await _addSharedSpaceMemberInteractor
          .execute(
              sharedSpaceId, AddSharedSpaceMemberRequest(accountId, sharedSpaceId, role.sharedSpaceRoleId))
          .then((result) => store.dispatch(SharedSpaceAction(result)));
      store.dispatch(SharedSpaceDetailsGetAllSharedSpaceMembersAction(
          await _getAllSharedSpaceMembersInteractor.execute(sharedSpaceId)));
    });
  }
```